### PR TITLE
feat: support ldap authenticate

### PR DESCRIPTION
**Pull request by csh0101**
support ldap authenticate,

- #293

### CODE REVIEW

**Comment by csh0101** [comment]
@dongliutony @SSebo  i m confused about this code analysis.. I think the logic is no method to reduce...

**Comment by gangb-tech** [comment]
Thank you for your pull request. I've reviewed the changes, and overall, it looks good. However, I have a few suggestions to enhance the PR before merging it.

Firstly, it would be great if you could associate this pull request with the relevant issue.  Please add a comment in the PR description or use the GitHub syntax to link the PR with the issue.

Secondly, it's essential to include corresponding unit tests for the changes you've made. 

Once you have made these updates, please let me know, and I'll be happy to take another look. Keep up the great work!

**Comment by csh0101** [comment]
> Thank you for your pull request. I've reviewed the changes, and overall, it looks good. However, I have a few suggestions to enhance the PR before merging it.感谢您的拉取请求。我已经审查了更改，总体上看起来不错。但是，在合并之前，我有一些建议来增强PR。
> 
> Firstly, it would be great if you could associate this pull request with the relevant issue. Please add a comment in the PR description or use the GitHub syntax to link the PR with the issue.首先，如果您能将此拉取请求与相关问题关联起来，那将是很好的。请在PR描述中添加评论或使用GitHub语法将PR与问题链接起来。
> 
> Secondly, it's essential to include corresponding unit tests for the changes you've made.其次，为您所做的更改包括相应的单元测试非常重要。
> 
> Once you have made these updates, please let me know, and I'll be happy to take another look. Keep up the great work!一旦您完成了这些更新，请告诉我，我很乐意再次查看。继续保持良好的工作！

Thanks I will do this.

**Comment by csh0101** [comment]
connects #293

**Comment by csh0101** [comment]
@gangb-tech  I add the support of unit test. 

But I think the test is do not more helpful for this. so Integration Test Maybe need.

and , some code I edited. 

Before, organization is created By Organization class's static method. 

when you call memberSerivce.save(),this is will trouble because of the member related a entity which not flush. In JPA Domain, it's  in a New Status. so it failed.

I remember I edit oraganizationRepo.findById to that in the last weekend. In my mind, I think they are equal with each other,but the truth is not.
It is a mercy that I test this again in my local environment.
it worker now! I commit a appliction.yaml with the ldap config.. 
please merge it.

**Comment by gangb-tech** [comment]
> @gangb-tech I add the support of unit test.
> 
> But I think the test is do not more helpful for this. so Integration Test Maybe need.
> 
> and , some code I edited.
> 
> Before, organization is created By Organization class's static method.
> 
> when you call memberSerivce.save(),this is will trouble because of the member related a entity which not flush. In JPA Domain, it's in a New Status. so it failed.
> 
> I remember I edit oraganizationRepo.findById to that in the last weekend. In my mind, I think they are equal with each other,but the truth is not. It is a mercy that I test this again in my local environment. it worker now! I commit a appliction.yaml with the ldap config.. please merge it.

Good job, but there are still some unresolved issues in the code review.

**Comment by csh0101** [comment]
oh,maybe you mean, a better view for the related issue. I do this. you can see the pr description.
now there are pr related issue which make sense

File: api/admin/src/test/groovy/io/featureprobe/api/service/UserPasswordAuthenticationProviderSpec.groovy [hunk 1]
```diff
@@ -41,34 +44,64 @@ class UserPasswordAuthenticationProviderSpec extends Specification {
...
+        1 * applicationContext.getBean(_) >> ldapAccountValidator
+        1 * applicationContext.getBean(_) >> new PlaintextEncryptionService()
+        1 * ldapTemplate.authenticate(_, _) >> { query,password ->
+            String uid = query.filter().toString().split("=")[1]
+            return uid == "Test" && password == "abc12345"
+        }
+        1 * memberRepository.findByAccount(account) >> Optional.of(new Member(account: account, password: "\$2a\$10\$jeJ25nROU8APkG2ixK6zyecwzIJ8oHz0ZNqBDiwMXcy9lo9S3YGma", status: MemberStatusEnum.ACTIVE,
+                organizationMembers: [new OrganizationMember(new Organization(id: 1, name: ""), new Member(), OrganizationRoleEnum.OWNER)]))
+        1 * operationLogRepository.save(_)
+        null != authenticate
```
Comment (csh0101) [inline api/admin/src/test/groovy/io/featureprobe/api/service/UserPasswordAuthenticationProviderSpec.groovy:94]: test

**Review by csh0101 — COMMENTED** [review]
done

**Review by gangb-tech — CHANGES_REQUESTED** [review]

**Comment by gangb-tech** [comment]
> I confrim there are nothing in code review,no comment. sorry, I cant see you comment in code review @gangb-tech

I'm sorry, it was my mistake. I didn't submit my comment.

File: api/admin/src/main/java/io/featureprobe/api/auth/LdapConfig.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.ldap.core.LdapTemplate;
+import org.springframework.ldap.core.support.LdapContextSource;
+
+@Configuration
+public class LdapConfig {
+
+    @Value("${app.security.ldap.url}")
```
[the remark this answers is not in the document]
Comment (csh0101) [reply api/admin/src/main/java/io/featureprobe/api/auth/LdapConfig.java]: resolved

**Comment by gangb-tech** [comment]
Great job! Additionally, there is another issue to consider: Is it more reasonable to allow existing users to have normal access when the LDAP service is unavailable?

**Comment by csh0101** [comment]
> 

this is my consideration for this situation.

if we want approve this, maybe there are two method.

1.  add finally block to include the rest code after both authenticate logic and catch exception. 
2. validator edit to validators, that support multiple validators. 

I dont confirm which broken the single resp design pattern. And I decide to choose the first one.

**Comment by csh0101** [comment]
> > 
> 
> this is my consideration for this situation.
> 
> if we want approve this, maybe there are two method.
> 
> 1. refix catch block, do not throw expection
> 2. validator edit to validators, that support multiple validators.
> 
> I dont confirm which broken the single resp design pattern. And I decide to choose the first one.

the mind is nod good, i will refix.

**Comment by csh0101** [comment]
@gangb-tech thanks, i add the logic for success login when a account exist in db on the situation that ldap server shutdown 
``` java 
try {
            ldapTemplate.authenticate(query().where("uid").is(token.getAccount()),token.getPassword());
        }
         catch (EmptyResultDataAccessException e) {
             throw new UsernameNotFoundException("Account not find on Ldap Server."+e.getMessage());
        } catch (AuthenticationCredentialsNotFoundException e) {
            throw new UsernameNotFoundException("Account password not correct on Ldap Server."+e.getMessage());
        } catch (Exception e) {
            log.error("err when authenticate on ldap server {}",e.getMessage());
        }
```

File: api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        } catch (AuthenticationCredentialsNotFoundException e) {
+            throw new UsernameNotFoundException("Account password not correct on Ldap Server."+e.getMessage());
+        } catch (Exception e) {
+            log.error("err when authenticate on ldap server {}",e.getMessage());
+        }
+        String memberName = token.getAccount();
+            Optional<Member> memberOptional =  memberService.findByAccount(memberName);
+            OrganizationMemberModel defaultOrganizeMemberModel =
+                    new OrganizationMemberModel(1L, "Default Organize", OrganizationRoleEnum.WRITER);
+            if (!memberOptional.isPresent()) {
```
Comment (gangb-tech) [inline api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java:73]: When the LDAP service is unavailable, account creation cannot be performed here.

**Review by gangb-tech — CHANGES_REQUESTED** [review]

File: api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.*;
+
+import static org.springframework.ldap.query.LdapQueryBuilder.query;
+
+
+@Component(value = "ldap")
+@AllArgsConstructor
+@Qualifier("ldap")
+@Slf4j
+@Primary
```
Comment (gangb-tech) [inline api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java:41]: remove @Primary

**Review by gangb-tech — CHANGES_REQUESTED** [review]

File: api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java (hunk 3 above)
Comment (csh0101) [reply api/admin/src/main/java/io/featureprobe/api/auth/LdapAccountValidator.java]: i add a flag check that is wether auth by ldap server.

**Comment by csh0101** [comment]
I refactor the method. to extract the authenticate by ldap in a single method.

just want to keep clean logic and clean code style.

**Review by gangb-tech — APPROVED** [review]

**Comment by gangb-tech** [comment]
LGTM

**Pull request merged**